src/http.cc

```diff
@@ -826,11 +839,113 @@ HttpStateData::handle1xx(HttpReply *reply)
     // for similar reasons without a 1xx response.
 }
 
+/// Matching protocols predicate function for std::find_if algorithm
+/// It handle the case the reference protocol includes a protocol version
+/// but the checking protocol does not (eg when squid requests upgrade to
+/// 'websockets' and server reports upgrade to 'websocket/1.2').
+class MatchProtocol
+{
+public:
+    explicit MatchProtocol(const char *s, size_t len) :ref(s), refLen(len) {}
+    bool operator() (const SBuf &checking) {
+        const char *p;
+        if (checking.find('/') == SBuf::npos && (p = std::find(ref, ref + refLen, '/')) != ref + refLen)
+            return checking.caseCmp(ref, p - ref) == 0;
+        else
+            return checking.caseCmp(ref, refLen) == 0;
+    }
+
+private:
+    const char *ref;
+    size_t refLen;
+};
+
+bool
+HttpStateData::upgradeProtocolsSupported(String &upgradeProtos)
+{
+    if (!upgradeProtocols)
+        return false;
+
+    const char *pos = NULL;
+    const char *item;
+    int ilen = 0;
+    while (strListGetItem(&upgradeProtos, ',', &item, &ilen, &pos)) {
+        auto it = std::find_if(upgradeProtocols->cbegin(), upgradeProtocols->cend(), MatchProtocol(item, ilen));
```

Comment:
Probably It is easy to do it. 
However I believe that it will result to more complex code. The MatchProtocol is enough big to be converted to a lamda-expression.

Reply:
>  it will result to more complex code.

I think that this would simplify and clarify the code instead.
For example, consider the following sketch:

```
SBuf checking(item, ilen);
auto it = std::find_if(upgradeProtocols->cbegin(), upgradeProtocols->cend(), [&checking] (const SBuf &p) {
      const auto checkingVersioned = checking.find('/') != SBuf::npos;
      const auto versioned = p.find('/') != SBuf::npos;
      return !versioned && checkingVersioned ? checking.startsWith(p) : (p == checking);
});
```
IMO, this sketch has two benefits:
* the algorithm becomses more clear because it showes when we need to search for the exact match and when just the prefix match.
* We do not need to maintain and document a special class, which is used only once.